django/core/handlers/asgi.py

```diff
@@ -0,0 +1,346 @@
+        self.resolver_match = None
+
+    @cached_property
+    def GET(self):
+        return http.QueryDict(self.scope.get("query_string", ""))
+
+    def _get_scheme(self):
+        return self.scope.get("scheme", "http")
+
+    def _get_post(self):
+        if not hasattr(self, "_post"):
+            self._read_started = False
+            self._load_post_and_files()
+        return self._post
+
+    def _set_post(self, post):
+        self._post = post
+
+    def _get_files(self):
+        if not hasattr(self, "_files"):
+            self._read_started = False
+            self._load_post_and_files()
+        return self._files
+
+    POST = property(_get_post, _set_post)
+    FILES = property(_get_files)
+
+    @cached_property
+    def COOKIES(self):
+        return http.parse_cookie(self.META.get("HTTP_COOKIE", ""))
+
+
+class AsgiHandler(base.BaseHandler):
+    """
+    Handler for ASGI requests.
+    """
+
+    request_class = AsgiRequest
+
+    # Size to chunk response bodies into for multiple response messages
+    chunk_size = 512 * 1024
+
+    def __init__(self):
+        super(AsgiHandler, self).__init__()
+        self.load_middleware()
+
+    async def __call__(self, scope, receive, send):
+        """
+        Async entrypoint - parses the request and hands off to get_response.
+        """
+        # Only serve HTTP connections (we should allow some way to override this)
+        if scope["type"] != "http":
+            raise ValueError(
+                "Django can only handle ASGI/HTTP connections, not %s"
+                % scope["type"]
+            )
+        # Receive the HTTP request body
+        body = await self.read_body(receive)
+        # If we've got here, then the request is complete and we can serve it.
+        set_script_prefix(self.get_script_prefix(scope))
+        # TODO: Is signal sending async-safe?
+        signals.request_started.send(sender=self.__class__, environ=scope)
+        # Get the request and check for basic issues
+        request, error_response = self.create_request(scope, body)
+        if request is None:
+            await self.send_response(error_response, send)
+            return
+        # Get the response, using a threadpool via sync_to_async if needed
+        if asyncio.iscoroutinefunction(self.get_response):
+            response = await self.get_response(request)
+        else:
+            # If get_response is synchronous, run it non-blocking
+            response = await sync_to_async(self.get_response)(request)
+        response._handler_class = self.__class__
+        # Increase chunk size on file responses (ASGI servers will handle low-level chunking)
+        if isinstance(response, FileResponse):
+            response.block_size = self.chunk_size
+        # Send the response
+        await self.send_response(response, send)
+
+    async def read_body(self, receive):
+        """
+        Reads a HTTP body from an ASGI connection.
+        """
+        body = b""
```

Comment:
Here you have a choice between `bytes` and `bytearray`. I ran this crude benchmark (Python 3.7, Linux, glibc):

```python
import time

CHUNK = b'\x00' * 1000
NUM_ITERS = 1000000

print('num_chunks,bytes_time,bytearray_time')
for num_chunks in range(1, 21):
    start = time.monotonic()
    for i in range(NUM_ITERS):
        body = b''
        for i in range(num_chunks):
            body += CHUNK
    finish = time.monotonic()
    bytes_time = time.monotonic() - start

    start = time.monotonic()
    for i in range(NUM_ITERS):
        body = bytearray()
        for i in range(num_chunks):
            body += CHUNK
        body = bytes(body)
    bytearray_time = time.monotonic() - start
    print(f'{num_chunks},{bytes_time},{bytearray_time}')
```

The result is:

![bench](https://user-images.githubusercontent.com/1223550/56269664-d69ab900-60fc-11e9-9393-ba4fd3998b87.png)

At least with the parameters I used, bytearray is slower for a small number of chunks, but looks linear, while bytes looks quadratic.

Reply:
Given this is likely going to be rewritten to use a temporary file in future, I will likely just leave it bytes for now.